src/mame/heathkit/tlb.cpp

```diff
@@ -1390,6 +1417,269 @@ ioport_constructor heath_gp19_tlb_device::device_input_ports() const
+		[this](offs_t offset, u8 &data, u8 mem_mask) { if (!machine().side_effects_disabled()) { tap_8000h(); } });
+
+	m_alphanumeric_mode_active = true;
+	m_graphics_mode_active = false;
+
+	m_allow_tlb_interrupts = true;
+	m_allow_imaginator_interrupts = false;
+	m_im2_d1_val = 1;
+
+	m_hsync_irq_raised = false;
+
+	m_mem_map = 1;
+
+	m_mem_view.select(m_mem_map);
+}
+
+void heath_imaginator_tlb_device::device_reset()
+{
+	m_mem_map = 1;
+
+	m_mem_view.select(m_mem_map);
+}
+
+/*
+ *  Memory map for the I-100
+ *
+ *    0x0000-0x1fff    Program ROM (Heath TLB board)
+ *    0x2000-0x3fff    Graphics Command Processor(GCP) ROM (I-100)
+ *    0x4000-0x5fff    Scratchpad RAM (Heath TLB board)
+ *    0x6000-0x7fff    Graphics Scratchpad  R/W RAM (I-100)
+ *    0x8000-0xbfff    Graphics Display R/W RAM(I-100)
+ *    0xc000-0xdfff    Optional Graphics Command Processor memory (I-100)
+ *    0xe000-0xffff    Display Memory (Heath TLB board)
+ */
+void heath_imaginator_tlb_device::mem_map(address_map &map)
+{
+	map.unmap_value_high();
+
+	map(0x0000, 0x1fff).view(m_mem_view);
+
+	// H19 standard ROM
+	m_mem_view[0](0x0000, 0x1fff).rom().region("maincpu", 0).unmapw();
+
+	// GCP ROM mapped to 0x0000 on power-up/reset
+	m_mem_view[1](0x0000, 0x1fff).rom().region("maincpu", 0x2000).unmapw();
+
+	// Normal spot of the GCP ROM
+	map(0x2000, 0x3fff).rom();
+
+	// TLB Scratchpad
+	map(0x4000, 0x40ff).mirror(0x1f00).ram();
+
+	// GCP Scratchpad
+	map(0x6000, 0x607f).mirror(0x1f80).ram();
+
+	// Graphics RAM
+	map(0x8000, 0xbfff).ram().share(m_p_graphic_ram);
+
+	// optional Graphics Command Processor (GCP)
+	// map(0xc000, 0xdfff);
+
+	// Alphanumeric RAM
+	map(0xe000, 0xe7ff).mirror(0x1800).ram().share(m_p_videoram);
+
+}
+
+void heath_imaginator_tlb_device::tap_6000h()
+{
+	if (m_mem_map != 0)
+	{
+		m_mem_map = 0;
+		m_mem_view.select(m_mem_map);
+	}
+}
+
+void heath_imaginator_tlb_device::tap_8000h()
+{
+	if (!m_allow_tlb_interrupts)
+	{
+		allow_tlb_intr();
+
+		set_irq_line();
+	}
+}
+
+void heath_imaginator_tlb_device::allow_tlb_intr()
+{
+	m_allow_tlb_interrupts = true;
+	m_allow_imaginator_interrupts = false;
+	m_im2_d1_val = 1;
+}
+
+void heath_imaginator_tlb_device::allow_hsync_intr()
+{
+	m_allow_tlb_interrupts = false;
+	m_allow_imaginator_interrupts = true;
+	m_im2_d1_val = 0;
+}
+
+/**
+ * An input/output request (IORQ) is sent to the TERMINAL LOGIC board only when
+ * both IORQ (pin 20 of U7) AND A4 are low.
+ */
+void heath_imaginator_tlb_device::io_map(address_map &map)
+{
+	heath_tlb_device::io_map(map);
+
+	// interrupt routing
+	map(0x10, 0x10).select(0x08).mirror(0x07).w(FUNC(heath_imaginator_tlb_device::config_irq_w));
+
+	// display enabling (graphics or alphanumeric)
+	map(0x30, 0x30).select(0xc0).mirror(0x0f).w(FUNC(heath_imaginator_tlb_device::display_enable_w));
+
+	map(0x50, 0x50).mirror(0x0f); // Prevent mapping to TLB when a4 is high
+	map(0xd0, 0xd0).mirror(0x0f); // Prevent mapping to TLB when a4 is high
```

Comment:
@galibert does `map` and `mirror` with no handler make sense?  I thought this wouldn’t do anything.